### PR TITLE
NPS Survey: Initial skeleton of NPS Survey block

### PR DESCRIPTION
**Pull request by mattsherman**
This block is used to display a Net Promoter Score (NPS) survey to the user. This block will be used as part of a flow to collect feedback from the user on how likely they are to recommend WordPress.com to a friend/family/colleague.

*Note: NPS survey is currently still in development and not used in production yet. The following is still outstanding and will be added in future PRs:*

* Styling/UX (the current PR focuses on getting the code structure
	in place, not concerning ourselves with the styling or wording yet)
* Wrapping strings for translation (since design and styling isn't finalized yet,
	no need to translate likely-to-change wording)
* Redux actions and state usage (this will likely remove the need for the
	`onDismissed` handler prop)

To test:
- Go to `http://calypso.localhost:3000/devdocs/blocks/nps-survey`
- Try answering survey (select a score; click Finish; then Dismiss on the thank you screen)
- Try dismissing without answering survey (select a score, or not; Then click I'd rather not answer)

### CODE REVIEW

**Comment by mattsherman** [comment]
Putting this back in progress -- I'm going to add a feature flag to enable this only in `development`.

**Comment by mattsherman** [comment]
Ready for review again.

/cc @jonathansadowski @jblz @mjuhasz

File: client/blocks/nps-survey/docs/example.jsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * Internal dependencies
+ */
+import NpsSurvey from '../';
+import { isEnabled } from 'config';
+
+class NpsSurveyExample extends PureComponent {
+	constructor( props ) {
+		super( props );
+		this.state = {
+			wasDismissed: false,
```
Comment (jblz) [inline client/blocks/nps-survey/docs/example.jsx]: Is there any reason to not just have the initial state be an empty object here? We're not trying to reference anything that'd fatal below that I can tell.

File: client/blocks/nps-survey/index.jsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+class NpsSurvey extends Component {
+	static propTypes = {
+		onDismissed: PropTypes.func,
+		name: PropTypes.string
+	}
+
+	constructor( props ) {
+		super( props );
+		this.state = {
```
Comment (jblz) [inline client/blocks/nps-survey/index.jsx]: same as above -- an empty object would probably be ok here.

File: client/blocks/nps-survey/docs/example.jsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		};
+		this.handleDismissed = this.handleDismissed.bind( this );
+	}
+
+	handleDismissed( event ) {
+		this.setState( {
+			wasDismissed: true,
+			wasSubmitted: event.wasSubmitted,
+			surveyName: event.surveyName,
+			recommendationValue: event.recommendationValue
```
Comment (jblz) [inline client/blocks/nps-survey/docs/example.jsx]: It probably wouldn't hurt to do a `parseInt` here as well

File: client/blocks/nps-survey/index.jsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			isSubmitted: false
+		};
+		this.handleRecommendationSelectChange = this.handleRecommendationSelectChange.bind( this );
+		this.handleFinishClick = this.handleFinishClick.bind( this );
+		this.handleDismissClick = this.handleDismissClick.bind( this );
+	}
+
+	handleRecommendationSelectChange( newRecommendationValue ) {
+		this.setState( {
+			recommendationValue: newRecommendationValue
```
Comment (jblz) [inline client/blocks/nps-survey/index.jsx]: I'd do the `parseInt` here as opposed to `handleChange` function in the handler since this will ultimately live in a different part of the codebase (the reducer fn) and explicit casting should probably happen there.

File: client/blocks/nps-survey/index.jsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+		// TODO: fire Redux action and use Redux state tree to determine when
+		// survey has been submitted
+		this.setState( {
+			isSubmitting: true
+		} );
+
+		// simulate requestresponse time
+		setTimeout( () => {
+			this.setState( {
+				isSubmitted: true
```
Comment (jblz) [inline client/blocks/nps-survey/index.jsx]: I'd also set `isSubmitting` to false here & disable the buttons below when:
`this.state.isSubmitting || this.state.isSubmitted`

...but this is functionally equivalent

File: client/blocks/nps-survey/recommendation-select.jsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+				value={ value }
+				selected={ this.props.value === value }
+				disabled={ this.props.disabled }
+				onChange={ this.props.onChange }
+			/>
+		);
+	}
+
+	render() {
+		const values = range( 0, 11 );
```
Comment (jblz) [inline client/blocks/nps-survey/recommendation-select.jsx]: these-numbers-go-to-11

:)

File: client/blocks/nps-survey/docs/example.jsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+		} );
+	}
+
+	render() {
+		// Since this component is still in-progress and unstyled, we will hide it from
```
Comment (jblz) [inline client/blocks/nps-survey/docs/example.jsx]: I'd do the `isEnabled` check in `client/devdocs/design/blocks.jsx` instead of showing this message in production.

File: client/devdocs/design/blocks.jsx [hunk 8]
```diff
@@ -134,6 +135,7 @@ export default React.createClass( {
 					<DailyPostButton />
 					<PostLikes />
 					<ReaderFeaturedVideo />
+					<NpsSurvey />
```
Comment (jblz) [inline client/devdocs/design/blocks.jsx]: See [above](https://github.com/Automattic/wp-calypso/pull/12331/files#r107243023) -- I'd do the `isEnabled` check here.

File: client/blocks/nps-survey/docs/example.jsx (hunk 1 above)
Comment (mattsherman) [inline client/blocks/nps-survey/docs/example.jsx]: I like to explicitly initialize because I find it makes the code clearer. You can just look at the constructor to see what all the different state props are and what the initial value of them are.

File: client/blocks/nps-survey/docs/example.jsx (hunk 3 above)
Comment (mattsherman) [inline client/blocks/nps-survey/docs/example.jsx]: We are in control of the sending of the event, so I don't think that is necessary. We should guarantee when we send the event that it is an int.

File: client/blocks/nps-survey/index.jsx (hunk 5 above)
Comment (mattsherman) [inline client/blocks/nps-survey/index.jsx]: Good catch -- I'll update.

File: client/blocks/nps-survey/index.jsx (hunk 4 above)
Comment (mattsherman) [inline client/blocks/nps-survey/index.jsx]: I'd like to keep the parseInt as low in the code as possible (in `RecommendationOption`), allowing all the rest of the code to just assume it is dealing with ints.

File: client/blocks/nps-survey/docs/example.jsx (hunk 7 above)
Comment (mattsherman) [inline client/blocks/nps-survey/docs/example.jsx]: I went back and forth between doing the check here and in `blocks.jsx` as you suggest. I put it here to avoid 'cluttering' `blocks.jsx`, but since you suggested putting the check there, I think I will move it there -- it certainly is a bit clearer.

File: client/blocks/nps-survey/recommendation-select.jsx (hunk 6 above)
Comment (mattsherman) [inline client/blocks/nps-survey/recommendation-select.jsx]: 🎸

**Comment by mattsherman** [comment]
Thanks for the review @jblz! I have left responses to your comments and will address some of them. Let me know if you feel really strongly about any of the ones I plan on keeping as is.

File: client/blocks/nps-survey/index.jsx (hunk 4 above)
Comment (jblz) [inline client/blocks/nps-survey/index.jsx]: I guess we have different definitions of low :)

I consider the bit that does the writing of the state data to memory (what will get pulled out into the reducer) to be the lowest level code & would apply type guarantees there.

I don't feel strongly enough about it to block ya on it though.

File: client/blocks/nps-survey/docs/example.jsx (hunk 3 above)
Comment (mjuhasz) [inline client/blocks/nps-survey/docs/example.jsx]: We can guarantee data validity (do the type check) on the boundaries, and trust the internal data validity unless we have a good reason to be overly defensive (e.g. security).

File: client/blocks/nps-survey/docs/example.jsx (hunk 1 above)
Comment (mjuhasz) [inline client/blocks/nps-survey/docs/example.jsx]: A properly written constructor leaves the resulting object in a valid state. In this case it'd be valid without explicit initialization but I also like the overview of state props.

**Comment by mjuhasz** [comment]
I tried this branch, and skimmed through the code as well. It works as expected :+1:

**Comment by ehg** [comment]
I'm curious as to the rationale of putting this in `blocks`, this is where a UI mockup would be handy :)

**Comment by mattsherman** [comment]
I put it in `blocks` because it is a reusable component that will be "connected to state and have the ability to dispatch action" (from the blocks README).

For our v1, we are ultimately going to show this block in a dialog when we show the user the survey, but we envision the desire to experience with different ways of showing this survey in the future, so having this as a reusable, encapsulated block made sense.

Here is a UI mockup of entire flow we are planning:

http://wydzwf.axshare.com

File: client/blocks/nps-survey/docs/example.jsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	constructor( props ) {
+		super( props );
+		this.state = {
+			wasDismissed: false,
+			wasSubmitted: false,
+			surveyName: null,
+			recommendationValue: null,
+		};
+		this.handleDismissed = this.handleDismissed.bind( this );
+	}
```
Comment (mcsf) [inline client/blocks/nps-survey/docs/example.jsx]: Drive-by tip: with ES6+ instance properties, this `constructor` definition could be replaced like so:

```js
class NpsSurveyExample extends PureComponent {
  // static indicates a class property (i.e., cross-instance)
  static displayName = 'NpsSurvey';

  // instance prop eval'd upon instantiation
  state = {
    wasDismissed: false,
    // …
  };

  // same, but thanks to the fat arrow we get a bound method
  handleDismissed = ( event ) => {
    this.setState( … );
  }
```

Comment (jblz) [inline client/blocks/nps-survey/docs/example.jsx]: TIL.  Thanks, @mcsf!

**Comment by ehg** [comment]
> I put it in blocks because it is a reusable component that will be "connected to state and have the ability to dispatch action" (from the blocks README).

That seems reasonable, now that I've looked at the mockups. Thanks @mattsherman!

File: client/blocks/nps-survey/docs/example.jsx (hunk 9 above)
Comment (mattsherman) [inline client/blocks/nps-survey/docs/example.jsx]: Thanks for the ES6 tips @mcsf!
